Automattic/woocommerce-payments
# Try: decouple test and live mode customers

Fixes #

This PR is an attempt at validating and fixing the concerns raised in https://github.com/Automattic/woocommerce-payments/pull/899#discussion_r490269452. 

New customers and payment methods are created in the correct metadata according to the current test mode setting. The payment methods displayed in the checkout are filtered to match the customer and test mode setting, preventing customers from using unavailable payment methods.

For older customers, this PR attempts to add the customer present in the old `wcpay_customer_id` metadata to `wcpay_customer_id_test` or `wcpay_customer_id_live` based on the current test mode setting. While that works for most cases, it is possible that a customer was created in another mode and shoppers would get an error.

To prevent and fix that, we catch the `resource_missing` exception and move the customer to the right mode if the error message indicates that the customer exists in the other mode.

#### Changes proposed in this Pull Request

* Decouple test and live customers into two separate meta keys
* Filter a user's tokens based on the `customer_id` and `test_mode`
* Migrate old customers to the new meta keys
* Check for customers saved in the wrong environment when fetching their payment methods
* Check for customers saved in the wrong environment when adding new cards from My account



#### Testing instructions



There are a few scenarios that need to be tested:
1. Test customer created on master and migrated with this PR on test mode 
2. Test customer created on master and migrated with this PR on live mode 
3. Live customer created on master and migrated with this PR on live mode 
4. Live customer created on master and migrated with this PR on test mode

For each scenario we should:

*If you don't want to make a purchase, just add a payment method using the My account page. It should be enough to create the customer and check if the payment methods are migrated successfully. Don't forget to at least add something to the cart and go to the checkout page to see the shopper's payment methods.*

1. Checkout `master`
2. Login with a WC User
2. Make a purchase or add a payment method
3. Switch to this branch
4. Apply the test mode change (if applicable) and go to the checkout page
5. Assert that it doesn't break when the current mode is different than the one in which the customer was created
6. Assert that the customer's payment methods are displayed correctly. They should only be displayed in the mode they were created
    - If the migration happens in the same mode as the customer creation, the customer and their tokens should be automatically imported
    - If the migration happens in a different mode, then the customer and their tokens should be migrated the right mode. Unless a user already has a `customer_id` in the current mode, no payment methods should be displayed
7. If live/test mode were switched, make another purchase or add a payment method
8. Once a single WC User has both a test and live customer, go to the checkout page in test and in live modes
9. Assert that the cards displayed on checkout are the ones for the current mode

**To test the add cards from My account fix, don't load the checkout page before adding a new card, as that would trigger the migration**

-------------------
